### DISCUSSION TITLE
HDMI 4(input)x2(output) with hyperhdr @ 120HZ with rullz usb grabber

### DISCUSSION THREAD

**Discussion by reedy-au** [awawa-dev/HyperHDR#496] [Q&A] [answered]
I don't think this is a HyperHDR issue but before I spend more $ ... I will try to ask.

I recently purchased one of these HDMI MATRIX from   https://www.aliexpress.com/item/1005004685856434.html
This has 4 input ports and 2 output ports(outputB & outputA)

When I try use input from a Xbox series X @ 120HZ and output to outputB it send through the HDMI signal fine to the hdmi2.1 port on my TV. ALM,VRR,ARC all work OK.

When I try to output to outputA(this is a rullz MS2109 grabber) AS WELL AS outputB the HDMI MATRIX limits the output on both output ports to 60HZ. This works so far with HyperHDR but is not using the full potential of the TV and XSX @120HZ

The HDMI Matrix has a resolution downscaler which seems to work fine but I don't know if it downscales the FPS/refresh rate...

The question is, is it the limitation of the RULLZ USB (says it supports 4k@30 or 1080@60) that would be causing this OR is there something that I could tweak in HyperHDR?

I tried to contact the seller but there is a language barrier, I think they said the input HZ must be the same as the output HZ and the resolution can be automatically scaled down. I'm not sure though and I can't find a datasheet for either.

Sorry if the question is vague, that's the best I could explain it.  

I wonder has anyone else come across this issue?

Thank you.

**Comment by awawa-dev** [comment, marked as answer]
Hi
It's limitation of the grabber and despitethat  HDMI splitter/matrix can reduce the video size it can't change the framerate...so it defaults to highest framerate supported by all devices (MS2109 is the limiter here). You need e.g. Ezcap 320, Ezcap 321, Ezcap 331 or Elgato HD60x to handle 1080p120. More information on my blog (latency test and grabbers reviews).

**Comment by awawa-dev** [comment]
BTW MS2109 is HDMI 1.4b grabber.

**Comment by reedy-au** [comment]
oh man... I thought so. 

The elgato is way too expensive...

I'm looking at the ezcap 331 or the ezcap 321.  Your blog says the ezcap 321 is slightly better but it costs 50% more than the 331.

Do you think the ezcap 321 would make that much difference? I might as well pay the extra if it does.

Thanks again mate.

**Comment by awawa-dev** [comment]
Make an experiment before: try to disable HDR for 4k 120 if enabled and reduce color format to 8bit (preferable 4:2:0). These Aliexpress matrixes/splitters also need to be verified first despite what they claim in the advertising (48Gbps processing can be only on the paper or only for pass-through but not for the scaler).
I don't like the Ezcap 331 not because it performs poorly, but because it is an example of bad engineering using cheap materials, which causes the device to overheat. That's my main complaint (I experienced also poor USB 2.0 compatibility), it has similar performance to the Ezcap 321. There's also an MS2130 on the horizon but it's probably MS2109 on steroids (real USB 3.0 support, still HDMI 1.4)

**Comment by awawa-dev** [comment]
MS2130: https://github.com/YuzukiHD/YuzukiLOHCC-PRO , on Aliexpress MS2130 it's in $12-$16 range but without loop-out. I will test it later. I cant find if 1080p120 on the input is guaranteed.

**Comment by reedy-au** [comment]
I will try this thanks.

Re the aliexpress matrix. It's not bad but you are right it does advertise 48Gbps rate and 1200MHz transmission bandwidth BUT the manual it comes with says 40Gbps MAX and 1 GHz MAX. That being said it does work fine with a XSX which says 120HZ, VRR,ALM are all supported and it looks the same playing games when plugged directly in to the TV with no wierd artifacts etc.   But who knows...I only bought it because it was cheap and had decent reviews.

I asked one of the popular sellers if the MS2130 would support input of 1080@120Hz but they could say anything 😄 I see nothing in reviews about it unfortunately.

Overheating and crashing? Is no good as ambilight is in use here at least 6+ hours a day.

Maybe I will just go with the 321 plus it's got a loop-out.

Cheers

ouch if it overheats

**Comment by reedy-au** [comment]
> Make an experiment before: try to disable HDR for 4k 120 if enabled and reduce color format to 8bit (preferable 4:2:0). These 

I tried but no go.

Also I'd like to report back that this matrix says it supports ARC but it is VERY flakey. Seeing as XSX only has 1 xHDMI output and a TOSLINK(i think but that sucks) that kills my splitter to matrix upgrade.  ^&*(

I tested with HDMI CEC on on all devices and off on all devices and forced ARC on my reciever.  The TV and receiver say they're connected but there is no sound 90% of the time. It got to the stage where my receiver wouldn't accept sound from any source and needed a hard power cycle. First time that's happened...

I have just reverted back to my previous ambilight setup and it works perfect as usual.

Maybe someone else has had some luck? If not hope this helps someone trying to get HyperHDR ambilight on EVERYTHING haha.

I will say the aliexpress hdmi matrix unit works fantastic for video. For the price it's worth it, I also tested 4k120hz on BOTH outputs and that works fine to 2 sony 2.1 hdmi TV's.

Can't win em all.

Thanks again awawa for your fast replies and hints in the right direction. 👍

**Comment by marcoisabella** [comment]
Sorry to bring this topic back. But I am looking for a valid splitter to use on my HyperHDR, MS2130 grabber, in order to keep 4k, dolby and CEC. Here in the US Feintech and Ezcoo are not available (currently). 

Are you still using it? How is it? Does the CEC work fine?

**Reply by rorosaurus** [reply]
Did you find something that works?

**Reply by cr333d** [reply]
Not sure if anyone have tried this [https://www.aliexpress.us/item/3256804124019650.html?af=%7B%22c%22%3A%22Y0WlCZS6Ia53ojmw1703336436%22%7D&aff_fcid=cd5d43f3bf264b33865cc66103bee627-1703342851892-08946-_ompl6t6&aff_fsk=_ompl6t6&aff_platform=api-new-link-generate&sk=_ompl6t6&aff_trace_key=cd5d43f3bf264b33865cc66103bee627-1703342851892-08946-_ompl6t6&terminal_id=d17e5dc91e284c97913e561c88771137&afSmartRedirect=y&gatewayAdapt=glo2usa](url)

![image](https://github.com/awawa-dev/HyperHDR/assets/15855098/272110d1-a3bb-43ad-ac1e-33d61c28c800)

**Comment by kokoccc** [comment]
Hi @reedy-au, I've just faced the same problem, and I hoped to know if you eventually managed to pull it off :)

In my case, I have similar devices:
- HDMI 2.1 4x2 Matrix - https://aliexpress.ru/item/1005004729344581.html?sku_id=12000030256963429
- Rullz MS2109 USB 2.0 Capture Card - https://aliexpress.ru/item/4001044845208.html?sku_id=12000031634495056

When I connect MS2109 to `Out A`, my HDMI 2.1 TV is able to receive only 4K/60 signal.
As I understood, the matrix can't handle 120Hz and 60Hz at the same time, as it needs the same frequency on both devices.

So, were you able to get 4K/120 from `Out B` using MS2130 on `Out A`?
This problem took a lot from me, huge thanks in advance for the answer!

**Reply by kokoccc** [reply]
Got MS2130 this weekend. Planned to use multiple grabbers at home for different purposes, so I decided to order all three models from here: https://aliexpress.ru/item/1005004990708725.html (the same seller as awawa-dev's).

Unfortunately, none of them supported 1080@120 output as expected. 120hz input wasn't reported either, and the Matrix was still unable to provide 4k@120 through `Out B` when MS2130 was connected to `Out A`, as before.

Regarding firmware: it appears that in my case only the `Shape B` model had the new firmware (which suggested both YUV and MJPEG when connected to USB 3 RPi 5 port). Both `Shape A` and `Shape C` models suggested YUV without MJPEG when connected to the same port.

I attempted to flash a custom firmware from [this thread](https://github.com/awawa-dev/HyperHDR/discussions/729#discussioncomment-8126044) onto `Shape B` and `Shape C` models. I installed Go, built cli from Ubuntu terminal, made backups, flashed - everything went smoothly, no warnings, no errors. I decided not to touch `Shape A` just to be safe.

Results:
- the matrix now supports 4k@120 through `Out B` (YAY!), even with HDR and Dolby Vision, verified with Nvidia Shield, RTX 4080 and PS5
- when connected to USB 3 port, `Shape B` model only suggests YUV in HyperHDR
- I can't select 120Hz in HyperHDR, the max value is 60Hz for every resolution up to 1080p
- when HDR is enabled, the image received by HyperHDR is too dark, and the LED strip itself is also too dark. This seems to be a similar issue to the one [described](https://github.com/awawa-dev/HyperHDR/discussions/729#discussioncomment-8126042) with Ezcoo, but I have't managed to fix it in my case. Turning the power off and on didn’t help either

Overall, my main issue with 4k@120 has been resolved, big thanks to awawa-dev and mates for tests and firmware, you guys rock!

And I'm happy with the delay. It’s noticeable in some test videos or when the smoothing feature is turned on, but it doesn’t significantly affect gameplay or movie watching. And Dustin Farrell's Transient looks awesome, yeah.

The last two issues from the result list remain unsolved. If anyone has suggestions or thoughts, I’d be happy to hear them. Thanks!

**Reply by awawa-dev** [reply]
> None of them supported 1080@120 output as expected.

@kokoccc ms2130 cannot capture using 120Fps, 1080p120Hz input is only available for very low bitrate (colorspace) usually with an assistance of the HDMI splitter scaller.

> when HDR is enabled, the image received by HyperHDR is too dark, and the LED strip itself is also too dark. This seems to be a similar issue to the one https://github.com/awawa-dev/HyperHDR/discussions/729#discussioncomment-8126042 with Ezcoo, but I have't managed to fix it in my case. Turning the power off and on didn’t help either

You can try to calibrate LUT using our new calibration procedure (it's easier than before) and we will see if there can be a workaround: https://github.com/awawa-dev/HyperHDR/pull/896 Installers: https://github.com/awawa-dev/HyperHDR/actions/runs/10872736016 Please provide us calibration_captured_yuv.txt file after the calibration.

**Reply by kokoccc** [reply]
> You can try to calibrate LUT using our new calibration procedure (it's easier than before) and we will see if there can be a workaround: https://github.com/awawa-dev/HyperHDR/pull/896 Installers: https://github.com/awawa-dev/HyperHDR/actions/runs/10872736016 Please provide us calibration_captured_yuv.txt file after the calibration.

Thanks @awawa-dev, have just tried this method and it seems the picture has become much brighter! Here's the file [calibration_captured_yuv.txt](https://github.com/user-attachments/files/17016056/calibration_captured_yuv.txt)

Now I'm watching ambilight test videos, and I see some mismatching colors, so trying to adjust gamma values and other options. And who knows, maybe the reason is my lightgray wall behind TV hehe :)

**Reply by awawa-dev** [reply]
Thanks. The brightness/nits were detected correctly but something strange has happened to colors and the result is not so great. Do you have maybe some kind of tone mapping or color enhancement enabled in your player (which one do you use?)? Did you use yuv420 HDR test file?

**Reply by kokoccc** [reply]
I'm using Win 11 MPC-HC, without any additional color tweaks applied. The file is yuv420 HDR (low quality).

I found out the night light (blue filter) was turned on, so I turned it off and rerun calibration - not sure if there are any differences in the result, though: [calibration_captured_yuv.txt](https://github.com/user-attachments/files/17017326/calibration_captured_yuv.txt). HDR is still active.

<details>
<summary>Some HyperHDR screenshots, just in case</summary>

![image](https://github.com/user-attachments/assets/1107c813-e139-43f4-9246-dada4043280b)

![image](https://github.com/user-attachments/assets/5b522f3c-f013-420f-9109-5c529a4f2438)

</details>

**Reply by awawa-dev** [reply]
Now it's completely different story: the results are excellent 😄 

<details>
<summary>Results</summary>

```
2024-09-16T19:58:44.959Z [CALIBRATOR] (LutCalibrator.cpp:1027) --------------------------------- Actual PQ primaries for YUV coefs: FCC ---------------------------------
2024-09-16T19:58:44.961Z [CALIBRATOR] (LutCalibrator.cpp:1028) r: (0.651, 0.329) vs sRGB(0.640, 0.330) vs bt2020(0.708, 0.292)
2024-09-16T19:58:44.961Z [CALIBRATOR] (LutCalibrator.cpp:1029) g: (0.289, 0.628) vs sRGB(0.300, 0.600) vs bt2020(0.170, 0.797)
2024-09-16T19:58:44.962Z [CALIBRATOR] (LutCalibrator.cpp:1030) b: (0.144, 0.055) vs sRGB(0.150, 0.060) vs bt2020(0.131, 0.046)
2024-09-16T19:58:44.963Z [CALIBRATOR] (LutCalibrator.cpp:1031) w: (0.313, 0.329) vs sRGB(0.313, 0.329) vs bt2020(0.313, 0.329)
2024-09-16T19:58:44.966Z [CALIBRATOR] (LutCalibrator.cpp:821) Score: 571.376000
2024-09-16T19:58:44.966Z [CALIBRATOR] (LutCalibrator.cpp:822) Time: 16.191000
2024-09-16T19:58:44.966Z [CALIBRATOR] (LutCalibrator.cpp:824) Selected coef: FCC
2024-09-16T19:58:44.967Z [CALIBRATOR] (LutCalibrator.cpp:825) Selected coef delta: -0.018000 -0.014000
2024-09-16T19:58:44.967Z [CALIBRATOR] (LutCalibrator.cpp:826) Selected EOTF: PQ
2024-09-16T19:58:44.968Z [CALIBRATOR] (LutCalibrator.cpp:833) Selected nits: 239.811132
2024-09-16T19:58:44.968Z [CALIBRATOR] (LutCalibrator.cpp:835) Selected bt2020 range: 1
2024-09-16T19:58:44.968Z [CALIBRATOR] (LutCalibrator.cpp:836) Selected alt convert: 1
2024-09-16T19:58:44.969Z [CALIBRATOR] (LutCalibrator.cpp:837) Selected aspect: 0.980000 1.100000 1.100000
2024-09-16T19:58:44.980Z [CALIBRATOR] Write captured colors to: C:/Users/:-)/.hyperhdr/calibration_captured_yuv.txt

HDR10:
       White: [255 255 255] => [253 253 253] [corrected]
         Red: [255   0   0] => [244   0   0] [corrected]
       Green: [  0 255   0] => [  0 247   0] [corrected]
        Blue: [  0   0 255] => [  0   0 255] [corrected]
    UpperRed: [192   0   0] => [187   0   0] [corrected]
  UpperGreen: [  0 192   0] => [ 16 185   0] [corrected]
   UpperBlue: [  0   0 192] => [  0   0 201] [corrected]
   MiddleRed: [128   0   0] => [124   0   0] [corrected]
 MiddleGreen: [  0 128   0] => [  0 126   0] [corrected]
  MiddleBlue: [  0   0 128] => [  0   0 133] [corrected]
      LowRed: [ 64   0   0] => [ 62   0   0] [corrected]
    LowGreen: [  0  64   0] => [  0  66   0] [corrected]
     LowBlue: [  0   0  64] => [  0   0  67] [corrected]
      Yellow: [255 255   0] => [246 251   0] [corrected]
     Magenta: [255   0 255] => [241   9 255] [corrected]
        Cyan: [  0 255 255] => [  0 255 245] [corrected]

      Orange: [255 128   0] => [245 130   0] [corrected]
    LimeBlue: [  0 255 128] => [  0 254 124] [corrected]
        Pink: [255   0 128] => [242   0 121] [corrected]
     LimeRed: [128 255   0] => [130 252   0] [corrected]
       Azure: [  0 128 255] => [  0 133 255] [corrected]
      Violet: [128   0 255] => [124   0 255] [corrected]
     WHITE_0: [  0   0   0] => [  0   0   0] [corrected]
     WHITE_1: [ 16  16  16] => [ 15  15  15] [corrected]
     WHITE_2: [ 32  32  32] => [ 35  35  35] [corrected]
     WHITE_3: [ 48  48  48] => [ 51  51  51] [corrected]
     WHITE_4: [ 64  64  64] => [ 67  67  67] [corrected]
     WHITE_5: [ 80  80  80] => [ 82  82  82] [corrected]
     WHITE_6: [ 96  96  96] => [ 97  97  97] [corrected]
     WHITE_7: [112 112 112] => [114 114 114] [corrected]
     WHITE_8: [128 128 128] => [128 128 128] [corrected]
     WHITE_9: [144 144 144] => [145 145 145] [corrected]
    WHITE_10: [160 160 160] => [159 159 159] [corrected]
    WHITE_11: [176 176 176] => [177 177 177] [corrected]
    WHITE_12: [192 192 192] => [193 193 193] [corrected]
    WHITE_13: [208 208 208] => [211 211 211] [corrected]
    WHITE_14: [224 224 224] => [225 225 225] [corrected]
    WHITE_15: [240 240 240] => [241 241 241] [corrected]
    WHITE_16: [255 255 255] => [253 253 253] [corrected]


-------------------------------------------------------------------------------------------------
                                        Detailed results
-------------------------------------------------------------------------------------------------
[  0   0   0] => [  0   0   0]
[  0   0  64] => [  0   0  67]
[  0   0 128] => [  0   0 133]
[  0   0 192] => [  0   0 201]
[  0   0 255] => [  0   0 255]
[  0  64   0] => [  0  66   0]
[  0  64  64] => [  0  66  63]
[  0  64 128] => [  0  68 128]
[  0  64 192] => [  0  69 191]
[  0  64 255] => [  0  69 255]
[  0 128   0] => [  0 126   0]
[  0 128  64] => [  0 125  62]
[  0 128 128] => [  0 129 123]
[  0 128 192] => [  0 131 187]
[  0 128 255] => [  0 133 255]
[  0 192   0] => [ 16 185   0]
[  0 192  64] => [ 17 186  65]
[  0 192 128] => [ 30 188 126]
[  0 192 192] => [  0 194 184]
[  0 192 255] => [  0 193 245]
[  0 255   0] => [  0 247   0]
[  0 255  64] => [  0 250  50]

[  0 255 128] => [  0 254 124]
[  0 255 192] => [ 22 250 193]
[  0 255 255] => [  0 255 245]
[ 64   0   0] => [ 62   0   0]
[ 64   0  64] => [ 63   0  64]
[ 64   0 128] => [ 61   0 133]
[ 64   0 192] => [ 63   0 198]
[ 64   0 255] => [ 53   0 255]
[ 64  64   0] => [ 63  65   0]
[ 64  64  64] => [ 67  67  67]
[ 64  64 128] => [ 64  69 128]
[ 64  64 192] => [ 65  71 194]
[ 64  64 255] => [ 67  71 255]
[ 64 128   0] => [ 69 127   0]
[ 64 128  64] => [ 63 127  64]
[ 64 128 128] => [ 66 130 121]
[ 64 128 192] => [ 61 131 187]
[ 64 128 255] => [ 71 130 250]
[ 64 192   0] => [ 79 186   0]
[ 64 192  64] => [ 59 191  63]
[ 64 192 128] => [ 64 193 125]
[ 64 192 192] => [ 56 193 180]
[ 64 192 255] => [ 64 192 240]
[ 64 255   0] => [ 89 245   0]
[ 64 255  64] => [ 30 251  54]
[ 64 255 128] => [ 79 247 125]
[ 64 255 192] => [ 76 252 189]
[ 64 255 255] => [ 57 255 246]
[128   0   0] => [124   0   0]

[128   0  64] => [120   0  62]
[128   0 128] => [119   0 126]
[128   0 192] => [124   0 191]
[128   0 255] => [124   0 255]
[128  64   0] => [121  68   0]
[128  64  64] => [118  70  64]
[128  64 128] => [123  70 125]
[128  64 192] => [121  74 188]
[128  64 255] => [123  73 251]
[128 128   0] => [121 128   0]
[128 128  64] => [124 129  62]
[128 128 128] => [128 128 128]
[128 128 192] => [129 130 183]
[128 128 255] => [124 130 250]
[128 192   0] => [130 186   0]
[128 192  64] => [130 188  65]
[128 192 128] => [132 187 125]
[128 192 192] => [122 193 184]
[128 192 255] => [131 195 249]
[128 255   0] => [130 252   0]
[128 255  64] => [123 254  54]
[128 255 128] => [125 254 127]
[128 255 192] => [123 255 189]
[128 255 255] => [133 255 246]
[192   0   0] => [187   0   0]
[192   0  64] => [184   0  66]
[192   0 128] => [183   0 125]
[192   0 192] => [186   0 191]
[192   0 255] => [185   0 255]

[192  64   0] => [178  67   0]
[192  64  64] => [186  66  61]
[192  64 128] => [179  71 122]
[192  64 192] => [179  76 191]
[192  64 255] => [186  74 255]
[192 128   0] => [181 126   0]
[192 128  64] => [179 130  64]
[192 128 128] => [177 133 122]
[192 128 192] => [180 135 183]
[192 128 255] => [178 136 242]
[192 192   0] => [188 190   0]
[192 192  64] => [185 188  62]
[192 192 128] => [181 195 121]
[192 192 192] => [193 193 193]
[192 192 255] => [192 196 248]
[192 255   0] => [202 248   0]
[192 255  64] => [197 250  57]
[192 255 128] => [195 247 120]
[192 255 192] => [186 255 185]
[192 255 255] => [183 255 241]
[255   0   0] => [244   0   0]
[255   0  64] => [244   0  63]
[255   0 128] => [242   0 121]
[255   0 192] => [240   0 187]
[255   0 255] => [241   9 255]
[255  64   0] => [246  58   0]
[255  64  64] => [252  55  63]
[255  64 128] => [243  67 123]
[255  64 192] => [241  73 184]
[255  64 255] => [240  71 251]
[255 128   0] => [245 130   0]
[255 128  64] => [235 132  59]
[255 128 128] => [241 130 126]
[255 128 192] => [246 133 182]
[255 128 255] => [237 137 250]
[255 192   0] => [240 194   0]
[255 192  64] => [237 194  59]
[255 192 128] => [239 194 119]
[255 192 192] => [235 200 185]
[255 192 255] => [244 199 239]
[255 255   0] => [246 251   0]
[255 255  64] => [241 253  56]
[255 255 128] => [243 254 124]
[255 255 192] => [243 255 188]
[255 255 255] => [253 253 253]
```


</details>




I must add it to the manual. There is another tricky Windows11 setting but it should be not visible when HDR is enabled: https://github.com/awawa-dev/HyperHDR/discussions/924

**Reply by kokoccc** [reply]
Got it. Once again, thanks a lot for your help! I've also tried some videos with HDR turned off, and it seems everything's ok, with basic colors at least 😎
